doc/designs/ech-api.md

```diff
@@ -0,0 +1,713 @@
+connection. In addition to loading those values, the application can also
+indicate via `for_retry` which ECHConfig value(s) are to be included in the
+`retry_configs` fallback scheme defined by the ECH protocol.
+
+`OSSL_ECHSTORE_num_keys()` allows a server to see how many usable ECH private
+keys are currently in the store, and `OSSL_ECHSTORE_flush_keys()` allows a
+server to flush keys that are older than `age` seconds.  The general model is
+that a server can maintain an `OSSL_ECHSTORE` into which it periodically loads
+the "latest" set of keys, e.g. hourly, and also discards the keys that are too
+old, e.g. more than 3 hours old. This allows for more robust private key
+management even if public key distribution suffers temporary failures.
+
+The APIs the clients and servers can use to associate an `OSSL_ECHSTORE`
+with an `SSL_CTX` or `SSL` structure:
+
+```c
+int SSL_CTX_set1_echstore(SSL_CTX *ctx, OSSL_ECHSTORE *es);
+int SSL_set1_echstore(SSL *s, OSSL_ECHSTORE *es);
+```
+
+ECH will be enabled for the relevant `SSL_CTX` or `SSL` connection
+when these functions succeed. Any previously associated `OSSL_ECHSTORE`
+will be `OSSL_ECHSTORE_free()`ed.
+
+To access the `OSSL_ECHSTORE` associated with an `SSL_CTX` or
+`SSL` connection:
+
+```c
+OSSL_ECHSTORE *SSL_CTX_get1_echstore(const SSL_CTX *ctx);
+OSSL_ECHSTORE *SSL_get1_echstore(const SSL *s);
+```
+The resulting `OSSL_ECHSTORE` can be modified and then re-associated
+with an `SSL_CTX` or `SSL` connection.
+
+Finer-grained client control
+----------------------------
+
+Applications that need fine control over which ECHConfigList (from those
+available) will be used, can query an `OSSL_ECHSTORE`, retrieving information
+about the set of "singleton" ECHConfigList values available, and then, if
+desired, down-select to one of those, e.g., based on the `public_name` that
+will be used. This would enable a client that selects the server address to use
+based on IP address hints that can also be present in an HTTPS/SCVB resource
+record to ensure that the correct matching ECH public value is used. The
+information is presented to the caller using the `OSSL_ECH_INFO` type, which
+provides a simplified view of ECH data, but where each element of an array
+corresponds to exactly one ECH public value and set of names.
+
+```c
+/*
+ * Application-visible form of ECH information from the DNS, from config
+ * files, or from earlier API calls. APIs produce/process an array of these.
+ */
+typedef struct ossl_ech_info_st {
+    int index; /* externally re-usable reference to this value */
+    char *public_name; /* public_name from API or ECHConfig */
+    char *inner_name; /* server-name (for inner CH if doing ECH) */
+    unsigned char *outer_alpns; /* outer ALPN string */
+    int outer_alpns_len;
+    unsigned char *inner_alpns; /* inner ALPN string */
+    int inner_alpns_len;
+    char *echconfig; /* a JSON-like version of the associated ECHConfig */
+} OSSL_ECH_INFO;
```

Comment:
Yep, it's intended to be public

The JSON-like thing has been useful for tracing/logging, and for display (in haproxy via it's "stats" socket thing). Basically it could be anything more human-readable than base64 encoded text that would work for an ECHConfig (currently supported, or including something not currently supported such as extensions).

Reply:
"never... in future" is always hard I guess - the main thing needed is to provide a way for application clients that don't know how to decode an ECHConfigList to select one ECHConfig from the set read from DNS (esp. if those came from >1 HTTPS RR value) - the `public_name` isn't visible to the application otherwise. Secondary thing is for client/server logging etc.

Now, it is fair to say that we don't have experience with what'll really be needed here, (real servers with >1 HTTPS RR published aren't deployed for now) so one could argue that punting on this for now could be wiser. OTOH, the "multi CDN" use-case has always been considered important for ECH, which is when this'd start to be important.

How about we mark this with a "TODO(ECH): revisit this later" and see if the deployment landscape has evolved sufficient to make a better decision later on?

Reply:
Yeah, this sounds fair to me.

Reply:
did that